### ISSUE TITLE
Remove `defaultValue` and `defaultValue'` hacks

### ISSUE THREAD

**Issue by niazim3** [JacquesCarette/Drasil#512]
As explained in issue #356 by @JacquesCarette, these hacks may prevent extending things and were likely put in place to accomodate for how some languages must have a value for all variables when they get created. 

Further mentioned: "the reason I call it a hack is that it is based on the type (in this case StateType) to have a "default value". Rather, the backend code generator should insist that all values have a reasVal instead. That way the "default" value can be application-specific rather than type-generic."

`defaultValue` and `defaultValue'` are defined in [code/Language/Drasil/Code/Imperative/AST.hs](https://github.com/JacquesCarette/Drasil/blob/cad5e614b0cfa68ce05f69ef7944bdfa53df77a6/code/Language/Drasil/Code/Imperative/AST.hs#L253-L265)

**Comment by bmaclach** [comment]
Note that `defaultValue` and `defaultValue'` have been replaced with type-specific `default` functions in finally tagless GOOL, but these are still hacks and should still be removed:

https://github.com/JacquesCarette/Drasil/blob/b72f7c1627f89acd04ba0ff1b3cdc06d9510270f/code/drasil-code/Language/Drasil/Code/Imperative/New.hs#L144-L148

From @JacquesCarette : "The best way to do this is to insist that all variables get initialized."

I'll note that while using a `reasVal` for a default value should work, there is one other place where the `default` functions are used: GOOL offers a `listExtend` function which extends a list by filling in the extended piece with the default value for that type.

**Comment by JacquesCarette** [comment]
Definitely a solid improvement.  We should see if any of those `default*` routines are used currently, and remove the ones that aren't. Is `listExtend` used? If not, remove it!

**Comment by bmaclach** [comment]
@JacquesCarette The `default*` routines are all used when initializing input variables. However, if we use your suggestion of using the typical value as the default value, then those uses of `default*` can be removed.

As for `listExtend`, it is unfortunately used when reading input files. And the typical value solution won't always work in this case. For example, when reading the TSD.txt and SDF.txt files in GlassBR, the values we are reading do not each have a typical value - these values don't even exist in Drasil (though perhaps they should, and then maybe the typical value solution could work)

**Comment by JacquesCarette** [comment]
It would be best if we could remove all 'dummy' initialisations -- whether through default or typical.
This is probably a deeper bug in the design of GOOL, where we are probably being too naive in places. Likely the only way to figure it out will be to trace every single use of `default*` routines and see what we "actually meant to say".
Most likely, what's going on is that we're jumping to trying to emit code prematurely and/or we're emitting code without having some necessary information in hand, because that information isn't available (anymore).

**Comment by bmaclach** [comment]
The only uses of `default*` are those described above: Initializing the input variables with values, and in `listExtend`, which itself is only used when reading input files. Looking deeper at each use:

Initializing input values: We can fix this by simply not doing it. I just tested removing this code from the generated GlassBR code for each of the 4 languages we currently generate, and nothing broke.

Extending lists while reading files: Instead of extending a list with default values to get it to be the correct size, we can just append the values to the list as they are encountered in the file being read. I tested this on a couple of the languages, too, and there were no problems.

**Comment by JacquesCarette** [comment]
I think Objective C requires all variables be initialized. Not that we care about Objective C that much (anymore), but other languages might come up that need it.

There was a language, I don't remember which, that required this "extend and initialize".

**Comment by bmaclach** [comment]
If a language requires all variables be initialized, the only way I can think of getting around dummy initializations is if we read the input right when the variables get declared, which would mean combining the input format and input parameters modules, which does not seem great.

I took a look at our old renderers for `ListExtend`. It was not even implemented in Objective C or Lua, and all of the other languages implement it as "`ListAppend` with the default value", so I don't foresee any problems with replacing any uses of `listExtend` with `listAppend` on the correct value. If we ever encountered a language that did not allow this, we could maybe declare the list with the correct size, thus avoiding any need to extend or append to the list. The information to determine the required size would be available (indeed, we are already using it to extend the list to the required size).

**Comment by JacquesCarette** [comment]
Re: initializing variables. I don't think it requires merging those modules. I think it requires inverting the control flow -- first values are read in, and then the 'final destination' for those values is created and initialized, instead of the other way around.

If `listExtend` is that much of a hack, maybe we can remove it now?

**Comment by bmaclach** [comment]
I'm all for removing `listExtend`. I can get to that shortly.

Even if we read in the values before creating/initializing the input parameters data structure, wouldn't we still need dummy initializations for the variables that will hold the values prior to assigning them to the 'final destination' variables? Unless the language is like Python and would allow the input format function to return multiple outputs that can be directly assigned to new variables.

**Comment by JacquesCarette** [comment]
Rather than theorize about this, we should try to figure out when this was implemented and why. "git blame" is usually a good starting point.

**Comment by bmaclach** [comment]
The dummy initializations were implemented with the very first commit of the code generator for the input module: 976f7bf03b2baca1ee2858592323799578dffb2f

It seems like this was done because the manual versions of the code, written previously, had these dummy initializations. I tried to trace that back and I could get as far back as here: 36f354380193320b5ad49bc7f478c90750aa26a3, which was the initial commit of GlassBR code to this repo, and the dummy initializations were already there. But it seems like prior to that the code existed on a Drasil repo in GitLab, which either no longer exists or I no longer have access to, so I couldn't check the history there.
